rust/sealevel/programs/mailbox/src/processor.rs

```diff
@@ -0,0 +1,855 @@
+        return Err(ProgramError::from(Error::DuplicateMessage));
+    }
+    inbox.delivered.insert(id);
+
+    let ism_ixn = IsmInstruction {
+        metadata: process.metadata,
+        message: message.body.clone(),
+    };
+    if inbox.ism != *next_account_info(accounts_iter)?.key {
+        return Err(ProgramError::from(Error::AccountOutOfOrder));
+    }
+    let mut ism_accounts = vec![auth_account.clone()];
+    let mut ism_account_metas = vec![AccountMeta {
+        pubkey: *auth_account.key,
+        is_signer: true,
+        is_writable: false,
+    }];
+    for pubkey in &inbox.ism_accounts {
+        let meta = AccountMeta {
+            pubkey: *pubkey,
+            // TODO this should probably be provided up front...
+            is_signer: false,
+            is_writable: false,
+        };
+        let info = next_account_info(accounts_iter)?;
+        if info.key != pubkey {
+            return Err(ProgramError::from(Error::AccountOutOfOrder));
+        }
+        ism_accounts.push(info.clone());
+        ism_account_metas.push(meta);
+    }
+    let verify = Instruction::new_with_borsh(inbox.ism, &ism_ixn, ism_account_metas);
+
+    let recp_ixn = RecipientInstruction {
+        sender: message.sender,
+        origin: message.origin,
+        message: message.body,
+    };
+    let recp_prog_id = Pubkey::new_from_array(message.recipient.0);
+    if recp_prog_id != *next_account_info(accounts_iter)?.key {
+        return Err(ProgramError::from(Error::AccountOutOfOrder));
+    }
+    let mut recp_accounts = vec![auth_account.clone()];
+    let mut recp_account_metas = vec![AccountMeta {
+        pubkey: *auth_account.key,
+        is_signer: true,
+        is_writable: false,
+    }];
+    for account in [] {
+        // TODO recipient accounts must be provided up front
+        let pubkey = Pubkey::new_from_array(account);
+        let meta = AccountMeta {
+            pubkey,
+            // TODO this should probably be provided up front...
+            is_signer: false,
+            is_writable: false,
+        };
+        let info = next_account_info(accounts_iter)?;
+        if info.key != &pubkey {
+            return Err(ProgramError::from(Error::AccountOutOfOrder));
+        }
+        recp_accounts.push(info.clone());
+        recp_account_metas.push(meta);
+    }
+    let recieve =
+        Instruction::new_with_borsh(message.recipient.0.into(), &recp_ixn, recp_account_metas);
+
+    if accounts_iter.next().is_some() {
+        return Err(ProgramError::from(Error::ExtraneousAccount));
+    }
+    let auth_seeds: &[&[u8]] = &[
+        b"hyperlane",
+        b"-",
+        &inbox.local_domain.to_le_bytes(),
+        b"-",
+        b"authority",
+        &[inbox.auth_bump_seed],
+    ];
+    invoke_signed(&verify, &ism_accounts, &[auth_seeds])?;
+    invoke_signed(&recieve, &recp_accounts, &[auth_seeds])?;
+    msg!("Hyperlane inbox: {:?}", id);
+
+    InboxAccount::from(inbox).store(inbox_account, true)?;
```

Comment:
I would argue the opposite in fact. If we recurse prior to storing, then we will burn through the whole compute budget rather than error out with something like message already delivered.